### PR TITLE
service/gsp: Implement saving of framebuffers in SaveVramSysArea.

### PR DESCRIPTION
**Pull request by Steveice10**
* Implements saving the current areas to specific VRAM locations in SaveVramSysArea.
* Updates the APT framebuffer capture for system applets to use these VRAM locations for capturing.
* Adds some extra safety checks to prevent crashes and other framebuffer capture bugs.

These changes allow the Amiibo Settings applet to launch from within games. The way captures previously worked would try to access the framebuffer info too late, and thus would get empty data. By implementing the copy to VRAM in SaveVramSysArea and using that, we now save the framebuffers at the right time and can copy them properly.

There are still some graphical and audio bugs that result from the context switching that need to be looked into separately.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6821)

### CODE REVIEW

**Review by FearlessTobi — APPROVED** [review]

File: src/core/hle/service/gsp/gsp_gpu.h [hunk 1]
```diff
@@ -186,6 +186,17 @@ struct CommandBuffer {
 };
 static_assert(sizeof(CommandBuffer) == 0x200, "CommandBuffer struct has incorrect size");
 
+static constexpr auto FRAMEBUFFER_WIDTH = 240;
```
Comment (GPUCode) [inline src/core/hle/service/gsp/gsp_gpu.h:189]: Global constants should have an explicit type. Also I think static is superfluous in this case, not too sure on that though

File: src/core/hle/service/apt/applet_manager.cpp [hunk 2]
```diff
@@ -1201,46 +1201,56 @@ void AppletManager::EnsureHomeMenuLoaded() {
...
-    static constexpr auto screen_width = 240;
-    static constexpr auto screen_width_pow2 = 256;
+    if (format == UINT32_MAX) {
+        // This format is used to indicate the framebuffer should not be copied.
+        return;
+    }
+
+    // Address in VRAM that APT copies framebuffer captures to.
+    static constexpr auto screen_capture_base_vaddr = Memory::VRAM_VADDR + 0x500000;
     const auto bpp = format < 2 ? 3 : 2;
```
Comment (GPUCode) [inline src/core/hle/service/apt/applet_manager.cpp:1211]: Format can be made into an enum for more readability https://www.3dbrew.org/wiki/NS_and_APT_Services#DisplayBufferMode

File: src/core/hle/service/apt/applet_manager.cpp [hunk 3]
```diff
@@ -1201,46 +1201,56 @@ void AppletManager::EnsureHomeMenuLoaded() {
...
+        return;
+    }
+
     for (u32 y = 0; y < height; y++) {
-        for (u32 x = 0; x < screen_width; x++) {
-            auto dst_offset =
-                VideoCore::GetMortonOffset(x, y, bpp) + (y & ~7) * screen_width_pow2 * bpp;
-            auto src_offset = bpp * (screen_width * y + x);
+        for (u32 x = 0; x < GSP::FRAMEBUFFER_WIDTH; x++) {
+            auto dst_offset = VideoCore::GetMortonOffset(x, y, bpp) +
```
Comment (GPUCode) [inline src/core/hle/service/apt/applet_manager.cpp:1233]:
```suggestion
            const auto dst_offset = VideoCore::GetMortonOffset(x, y, bpp) +
```
ditto for some other small const correctness stuff

File: src/core/hle/service/gsp/gsp_gpu.cpp [hunk 4]
```diff
@@ -704,18 +704,73 @@ void GSP_GPU::ImportDisplayCaptureInfo(Kernel::HLERequestContext& ctx) {
...
+
+    Memory::RasterizerFlushVirtualRegion(src, stride * lines, Memory::FlushMode::Flush);
+    std::memcpy(dst_ptr, src_ptr, stride * lines);
+    Memory::RasterizerFlushVirtualRegion(dst, stride * lines, Memory::FlushMode::Invalidate);
+}
+
 void GSP_GPU::SaveVramSysArea(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp(ctx);
 
+    if (active_thread_id == UINT32_MAX) {
```
Comment (GPUCode) [inline src/core/hle/service/gsp/gsp_gpu.cpp:724]: ditto for numeric_limits

**Review by GPUCode — APPROVED** [review]

File: src/core/hle/service/apt/applet_manager.cpp [hunk 5]
```diff
@@ -1199,21 +1199,37 @@ void AppletManager::EnsureHomeMenuLoaded() {
     }
 }
 
+static u32 GetDisplayBufferModePixelSize(DisplayBufferMode mode) {
+    switch (mode) {
+    // NOTE: APT does in fact use pixel size 3 for R8G8B8A8 captures.
```
Comment (GPUCode) [inline src/core/hle/service/apt/applet_manager.cpp:1204]: Does this need special logic for rasterizer cache flushing? If the owner surface is RGBA8 RasterizerFlushVirtualRegion will not flush the entire region correctly when called with bpp 3

Comment (Steveice10) [reply src/core/hle/service/apt/applet_manager.cpp:1211]: I don't know. I am just flushing/invalidating the exact memory regions that the framebuffer is copied from and to.

**Pull request merged**